Glass Refactor and other assorted minor structure check/tooltip fixes

Ruling-0 opened the pull request:
A glass refactor, and taking the time to hit a lot of low hanging fruit with structure checks and tooltips while boosting consistency.

Highlight: Refactors glass to all use one single structure element, which contains all (tiered) glass. Now, all multis that used ofBoroGlass, ofGlassTiered, ofGlassTieredMixed, or chainAllGlass, now all use chainAllGlass, which can be used as tiered or untiered.

**25 Mar**: glassTier is now a protected variable in MTEMultiBlockBase. Any multi can thus use chainAllGlasses() and just read glassTier as-needed. MTEMultiBlockBase.checkStructure handles resetting glassTier to -1. This method is only overridden in MTELargerTurbineBase (which has no glass); if any new multi overrides this and has glass it will need to either call super or handle it itself. Since every tiered-glass multi didn't do anything special with the getter/setter besides change the value of the local glassTier variable, they now all use chainAllGlasses() without arguments.

All multis using this element use the same glass subchannel for structure hologram projector, with the same indices - so tier 8 glass will always use Cosmic Neutronium Borosilicate on any multi that uses the element. The tiering is set up as follows
| 1-10 for HV-UMV | Variants, tier increasing | UMV |

UMV glass put at the end so full-stack hologram projector still places maximum tier glass. This relies on a very small PR to BlockRenderer6343: https://github.com/GTNewHorizons/BlockRenderer6343/pull/27

Small change to coremod since logic there obsoleted: https://github.com/GTNewHorizons/NewHorizonsCoreMod/pull/1216

Deprecates the BW API since addons very likely use it, but all usage internally in GT5 should be dead. Migrated GlassTier, RegisterGlassTiers, and GlassTooltipHandler from bartworks to GT API as GlassTier - this is now the basis of glass tiering. To cut down on bloat, the Railcraft glasses no longer qualify as HV-tier glasses.

Another notable change: frame boxes are now registered as machine blocks that can propagate structure updates, since many multis are starting to use them. This, combined with a minor structure update, also prevents around 95% of the autoclave multi from being technically optional. The update adds 8 extra casings; you can see 4 of them behind the glass in the below image, and there are a matching 4 in the back. Fixes https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/18973

Other minor changes:
- Add Bifrost glass as IV-tiered glass, closes https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/19070
- Add Chemical Grade glass as EV-tiered, Electron-Permeable and Infinity-Infused as UHV, Non-Photonic Matter Exclusion as UEV, and Hawking Radiation Alignment as UIV
- Add check to CoAL structure check that the casing is valid, fixes https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/19071
- Add check to QFT structure check that the shielding/manipulators are valid, fixes https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/17829
- Added more channel highlighting to structure tooltips, where it states the subchannels the multi can use and what they affect. 
- Added notice to AAL tooltip that recipes are limited to the energy hatch, closes https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/18163
- Making references to glass consistent in tooltips, closes https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/17852
- Moved glass registration and tooltip generation also into GT
- Registered the BHC and waterline glasses as tiered
- Added numbers to the tooltips of coils/glass/solenoids for easier tiering, closes https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/17166
- Removed Cyclotron tooltip about IV+ tier hatches, fixes https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/16916
- Cleaned up the Cyclotron file of many redundant overrides and removed muffler because it's similarly nonsensical
- Added standardized/helpful tooltip macros for laser-limiting or energy hatch-limiting glass tiering - these are respectively addMinGlassForLaser and addGlassEnergyLimitInfo
- All channel names which used spaces now use underscores (following Discord poll)

Some images:
![image](https://github.com/user-attachments/assets/621fc726-46a9-43cc-b249-2affb584aa3b)
![image](https://github.com/user-attachments/assets/04ffc8ff-1a05-47b7-9903-c928a6e47bc0)
![image](https://github.com/user-attachments/assets/40f91cdc-2ba9-4686-abbc-64b0c2455012)

fixes https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/19089
closes https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/18152
@jurrejelle contributions to fully remove cyclotron muffler, fixes https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/19394

Lyfts reviewed (commented):
Does every single type of glass really need its own individual tier? 40 tiers of glass when only 13ish of them are relevant seems excessive.

Ruling-0 commented:
> Does every single type of glass really need its own individual tier? 40 tiers of glass when only 13ish of them are relevant seems excessive.

If we have the ability to use them for multis, why not let them be autoplaced? We could definitely cull some of them (like the painted borosilicate). I already removed the _16_ railcraft HV-tier glasses.

Lyfts commented:
They can already be auto placed. My question is why do they each need their own individual instead of all of EV glasses sharing one, all IV glasses sharing and so on?

Ruling-0 commented:
> They can already be auto placed. My question is why do they each need their own individual instead of all of EV glasses sharing one, all IV glasses sharing and so on?

How can you specify which of them gets autoplaced out of a collection? Having to put all the blocks in inventory is quite cumbersome when just expanding the tier count is pretty straightforward and easy to use. With this way I can set projector to the tier I want (all now visible on tooltip, so I don't have to search the whole tier slider) and it will pull it from AE.

Lyfts commented:
Most of these glasses feel obscenely niche and I just don't think that warrants this kinda implementation. 
Whenever this gets merged we should 100% limit the amount of glasses shown in the nei preview to only the relevant ones. There's just no point in showing 37 different ones especially since the majority don't actually correspond to the structures tier anymore.

Ruling-0 commented:
> Most of these glasses feel obscenely niche and I just don't think that warrants this kinda implementation. Whenever this gets merged we should 100% limit the amount of glasses shown in the nei preview to only the relevant ones. There's just no point in showing 37 different ones especially since the majority don't actually correspond to the structures tier anymore.

I think limiting the NEI preview to the borosilicate 10, but keeping the range of tier values so they are specifically autoplaceable, is fair (maybe even an option to toggle the viewer?). We could also discuss if the variant tiered glass should be supported anyways, but so long as it is it should be specifiable somehow.

Glease commented on src/main/java/gtPlusPlus/xmod/gregtech/common/tileentities/machines/multi/production/MTECyclotron.java:
```diff
@@ -143,7 +132,7 @@ protected SoundResource getActivitySoundLoop() {
     public boolean checkMachine(IGregTechTileEntity aBaseMetaTileEntity, ItemStack aStack) {
         mCasing = 0;
-        return checkPiece(mName, 7, 1, 12) && mCasing >= 40 && checkHatch();
+        return checkPiece(mName, 7, 1, 12) && mCasing >= 40;
```
Why?

Ruling-0 replied:
In this case, because that function was not overridden so it was just True.

Glease commented on src/main/java/gtPlusPlus/xmod/gregtech/common/tileentities/machines/multi/production/MTEQuantumForceTransformer.java:
```diff
@@ -231,7 +231,7 @@ public boolean checkMachine(IGregTechTileEntity aBaseMetaTileEntity, ItemStack a
         }
 
-        if (mOutputBusses.isEmpty() || mOutputHatches.isEmpty()) {
+        if (this.mCraftingTier == -1 || this.mFocusingTier == -1) {
```
Why remove the hatch checks?

Glease commented on src/main/java/gregtech/common/tileentities/machines/multi/MTEMultiLathe.java:
```diff
@@ -248,7 +253,7 @@ public boolean checkMachine(IGregTechTileEntity aBaseMetaTileEntity, ItemStack a
         if (!checkPiece(STRUCTURE_PIECE_BODY, 3, 4, -1) && !checkPiece(STRUCTURE_PIECE_BODY_ALT, 3, 4, -1))
             return false;
-        return this.mMaintenanceHatches.size() == 1 && pipeTier > 0 && !mEnergyHatches.isEmpty() && mCasingAmount >= 42;
+        return pipeTier > 0 && !mEnergyHatches.isEmpty() && mCasingAmount >= 42;
```
Why remove mMaintenanceHatches check?

Glease commented on src/main/java/bartworks/common/tileentities/multis/MTEBioVat.java:
```diff
@@ -318,16 +317,12 @@ private boolean addRadiationInputToMachineList(IGregTechTileEntity aTileEntity,
-            && this.mMaintenanceHatches.size() == 1
-            && !this.mInputHatches.isEmpty()
-            && !this.mEnergyHatches.isEmpty();
+        return this.mCasing >= 19 && this.mRadHatches.size() <= 1 && !this.mEnergyHatches.isEmpty();
```
Again why do you remove so many hatch count checks?

Ruling-0 replied:
Consistency. I/O and Maint are very obvious if they are missing, because the thing doesn't work when built, and so most other multis do not check for them. This leads to some confusion since one multi will build, show valid structure (but has problems), while another will show invalid structure until the hatches are added. The energy hatch checks remain (and in the future probably needs its own pass) since it is less obvious when those are missing (they don't powerfail, they just never run).

The alternative would be adding hatch checks to all the other multis. I thought this was very minor - if it's more contentious, I'll revert and it can be discussed more in Discord.

Glease replied:
with void protection being the norm ye i'd say it's slightly less necessary to have these checks, but at least please keep the maint hatch count check in.

Glease reviewed (commented):
1. GT++ ChemPlant uses `pipe` channel for fluid pipe casing blocks. Might as well change that over
2. Based on the recent poll we should probably name channels using lower snake case, e.g. fluid_pipe
3. I'd prefer "Any Glass" over just "Glass", but this is a bit personal.

Sheodar commented:
Autoclave doesn't work (nightly 909)

Glease commented:
yeah also forget to mention that one, autoclave structure is now changed. You should probably mention it in a more explicit way than just saying "also prevents around 95% of the autoclave multi from being technically optional"

Ruling-0 commented:
Made the parent comment more explicit about autoclave structure. I saw and agree with the poll - I'll get the channel names converted. Additionally, I am aware of another bug with any multi not explicitly relying on glass tier - my untiered version of the call doesn't work - which I am working to address.

Ruling-0 commented:
- Addressed comments from @Glease
- Fixed the bug with some multis not allowing glass tier changes (and thus allowing their glass to be removed) - the most elegant solution involved making glassTier a protected attribute of MTEMultiBlockBase
- Changed the default tier for glass (and other things I touched) to -1 instead of -2 (I prefer it, it is more intuitive; it's PrAss casings that remain special)

jurrejelle commented:
https://github.com/GTNewHorizons/GT5-Unofficial/pull/4097 Please look at this PR, I made it thinking this PR was merged already :)

Dream-Master commented:
> #4097 Please look at this PR, I made it thinking this PR was merged already :)

Are we need to reopen the pr ? @jurrejelle

jurrejelle commented:
> > #4097 Please look at this PR, I made it thinking this PR was merged already :)
> 
> Are we need to reopen the pr ? @jurrejelle

Either Ruling-0 implements my changes into his branch, I create a new PR from my chagnes into this branch, or this PR gets merged and then mine gets merged. I'm equal on the first 2 options, wouldn't like the last ^^

Dream-Master commented:
> > > #4097 Please look at this PR, I made it thinking this PR was merged already :)
> > 
> > 
> > Are we need to reopen the pr ? @jurrejelle
> 
> Either Ruling-0 implements my changes into his branch, I create a new PR from my chagnes into this branch, or this PR gets merged and then mine gets merged. I'm equal on the first 2 options, wouldn't like the last ^^

whatever works

Ruling-0 commented:
After discussion with @Glease on Discord I have learned that there was a much better way around the untiered-structure issue than adding to MTEMultiBlock base - returning null for the tier of any non-tiered glass block. With this change I reverted back to the previous system but the bug of forming without glass is fixed. I have also expanded the scope slightly to quickly fix all the other instances of ofBlocksTiered to use null for invalid blocks instead of {0, -1, -2, -3}, so that now it is all to that standard (so anyone building new blocks, borrowing old code, will use it and leverage the early return it entails).

Additional changes in this last round of commits:
- Removed all the HV glass bloat (the painted borosilicate, and EIO bar Quite Clear Glass which is decently popular)
- Added the God Forge and Antimatter glasses as UMV glass alternates
Total glass tiers now 30 down from 39.

Glease reviewed (commented):
I see you reverted MBF changes completely in recent commits. Is this intentional?

Ruling-0 commented:
> I see you reverted MBF changes completely in recent commits. Is this intentional?

Good catch, I'm not sure how that happened. Fixed.

Glease reviewed (approved):
A job well done. Such kind of code polishing and massive bugfixes is exactly what we need 👍

YannickMG commented:
@Ruling-0 The changes to casing tooltips explaining which pipe channels to used were accidentally overwritten in the merge of https://github.com/GTNewHorizons/GT5-Unofficial/pull/4067

Those changes were incompatible with other changes to tooltips @RecursivePineapple was making, and those are the ones that are in master right now.

Those changes need to be reapplied in `ItemCasings`, hopefully in a way that can keep all the desired functionality.
